File: concepts/04-performance/node/batch-processing.js
/**
 * Batch Processing Implementation
 * Demonstrates batch operations for improved performance
 */

/**
 * Batch Processor
 */
class BatchProcessor {
  constructor(config = {}) {
    this.batchSize = config.batchSize || 10;
    this.flushInterval = config.flushInterval || 5000; // 5 seconds
    this.batch = [];
    this.processing = false;
    this.stats = {
      processed: 0,
      batches: 0,
      errors: 0
    };
    
    // Start flush timer
    this.startFlushTimer();
  }

  /**
   * Add item to batch
   */
  async add(item) {
    this.batch.push({
      item,
      timestamp: Date.now()
    });

    // Auto-flush if batch is full
    if (this.batch.length >= this.batchSize) {
      await this.flush();
    }
  }

  /**
   * Process batch
   */
  async processBatch(batch) {
    // Simulate batch processing
    console.log(`Processing batch of ${batch.length} items...`);
    
    const results = [];
    for (const entry of batch) {
      try {
        // Simulate processing
        await new Promise(resolve => setTimeout(resolve, 10));
        results.push({ success: true, item: entry.item });
        this.stats.processed++;
      } catch (error) {
        results.push({ success: false, item: entry.item, error });
        this.stats.errors++;
      }
    }
    
    return results;
  }

  /**
   * Flush current batch
   */
  async flush() {
    if (this.processing || this.batch.length === 0) {
      return;
    }

    this.processing = true;
    const batchToProcess = [...this.batch];
    this.batch = [];

    try {
      const results = await this.processBatch(batchToProcess);
      this.stats.batches++;
      return results;
    } finally {
      this.processing = false;
    }
  }

  /**
   * Start flush timer
   */
  startFlushTimer() {
    setInterval(() => {
      this.flush();
    }, this.flushInterval);
  }

  /**
   * Get statistics
   */
  getStats() {
    return {
      ...this.stats,
      pendingItems: this.batch.length,
      processing: this.processing
    };
  }

  /**
   * Shutdown and flush remaining
   */
  async shutdown() {
    // Stop flush timer
    clearInterval(this.flushTimer);
    
    // Flush remaining items
    while (this.batch.length > 0) {
      await this.flush();
      await new Promise(resolve => setTimeout(resolve, 100));
    }
  }
}

/**
 * Database Batch Operations
 */
class DatabaseBatchOperations {
  constructor() {
    this.insertBatch = [];
    this.updateBatch = [];
    this.deleteBatch = [];
  }

  /**
   * Batch insert
   */
  async batchInsert(items) {
    if (items.length === 0) return [];
    
    // Simulate batch INSERT
    console.log(`Batch inserting ${items.length} records...`);
    await new Promise(resolve => setTimeout(resolve, items.length * 5));
    
    return items.map((item, index) => ({
      id: Date.now() + index,
      ...item,
      inserted: true
    }));
  }

  /**
   * Batch update
   */
  async batchUpdate(updates) {
    if (updates.length === 0) return [];
    
    // Simulate batch UPDATE
    console.log(`Batch updating ${updates.length} records...`);
    await new Promise(resolve => setTimeout(resolve, updates.length * 3));
    
    return updates.map(update => ({
      ...update,
      updated: true,
      updatedAt: Date.now()
    }));
  }

  /**
   * Batch delete
   */
  async batchDelete(ids) {
    if (ids.length === 0) return [];
    
    // Simulate batch DELETE
    console.log(`Batch deleting ${ids.length} records...`);
    await new Promise(resolve => setTimeout(resolve, ids.length * 2));
    
    return ids.map(id => ({ id, deleted: true }));
  }
}

// Example usage
async function demonstrateBatchProcessing() {
  console.log('=== Batch Processing ===\n');

  // Batch processor
  const processor = new BatchProcessor({
    batchSize: 5,
    flushInterval: 3000
  });

  // Add items
  console.log('Adding items to batch...');
  for (let i = 1; i <= 12; i++) {
    await processor.add({ id: i, data: `item-${i}` });
  }

  // Wait a bit
  await new Promise(resolve => setTimeout(resolve, 1000));
  console.log('Stats:', processor.getStats());

  // Flush remaining
  await processor.flush();
  console.log('Final stats:', processor.getStats());

  // Database batch operations
  console.log('\n=== Database Batch Operations ===\n');
  const dbOps = new DatabaseBatchOperations();

  // Batch insert
  const items = Array.from({ length: 10 }, (_, i) => ({
    name: `Item ${i}`,
    value: i * 10
  }));
  const inserted = await dbOps.batchInsert(items);
  console.log(`Inserted ${inserted.length} records`);

  // Batch update
  const updates = inserted.slice(0, 5).map(item => ({
    id: item.id,
    name: `Updated ${item.name}`
  }));
  const updated = await dbOps.batchUpdate(updates);
  console.log(`Updated ${updated.length} records`);

  // Batch delete
  const ids = inserted.slice(5).map(item => item.id);
  const deleted = await dbOps.batchDelete(ids);
  console.log(`Deleted ${deleted.length} records`);
}

if (require.main === module) {
  demonstrateBatchProcessing();
}

module.exports = { BatchProcessor, DatabaseBatchOperations };

File: concepts/04-performance/python/batch_processing.py
"""
Batch Processing Implementation
Demonstrates batch operations for improved performance
"""
import asyncio
from typing import List, Dict, Any, Optional
from dataclasses import dataclass, field
from datetime import datetime


@dataclass
class BatchEntry:
    """Batch entry"""
    item: Any
    timestamp: float


class BatchProcessor:
    """Batch processor implementation"""
    
    def __init__(self, config: Dict[str, Any] = None):
        config = config or {}
        self.batch_size = config.get('batch_size', 10)
        self.flush_interval = config.get('flush_interval', 5.0)  # 5 seconds
        self.batch: List[BatchEntry] = []
        self.processing = False
        self.stats = {
            'processed': 0,
            'batches': 0,
            'errors': 0
        }
        
        self._flush_task = None
        self._start_flush_timer()
    
    def _start_flush_timer(self):
        """Start flush timer"""
        async def flush_loop():
            while True:
                await asyncio.sleep(self.flush_interval)
                await self.flush()
        
        self._flush_task = asyncio.create_task(flush_loop())
    
    async def add(self, item: Any):
        """Add item to batch"""
        self.batch.append(BatchEntry(
            item=item,
            timestamp=datetime.now().timestamp()
        ))
        
        # Auto-flush if batch is full
        if len(self.batch) >= self.batch_size:
            await self.flush()
    
    async def process_batch(self, batch: List[BatchEntry]) -> List[Dict[str, Any]]:
        """Process batch"""
        print(f"Processing batch of {len(batch)} items...")
        
        results = []
        for entry in batch:
            try:
                # Simulate processing
                await asyncio.sleep(0.01)
                results.append({'success': True, 'item': entry.item})
                self.stats['processed'] += 1
            except Exception as e:
                results.append({'success': False, 'item': entry.item, 'error': str(e)})
                self.stats['errors'] += 1
        
        return results
    
    async def flush(self) -> Optional[List[Dict[str, Any]]]:
        """Flush current batch"""
        if self.processing or not self.batch:
            return None
        
        self.processing = True
        batch_to_process = self.batch.copy()
        self.batch.clear()
        
        try:
            results = await self.process_batch(batch_to_process)
            self.stats['batches'] += 1
            return results
        finally:
            self.processing = False
    
    def get_stats(self) -> Dict[str, Any]:
        """Get statistics"""
        return {
            **self.stats,
            'pending_items': len(self.batch),
            'processing': self.processing
        }
    
    async def shutdown(self):
        """Shutdown and flush remaining"""
        if self._flush_task:
            self._flush_task.cancel()
        
        # Flush remaining items
        while self.batch:
            await self.flush()
            await asyncio.sleep(0.1)


class DatabaseBatchOperations:
    """Database batch operations implementation"""
    
    async def batch_insert(self, items: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
        """Batch insert"""
        if not items:
            return []
        
        print(f"Batch inserting {len(items)} records...")
        await asyncio.sleep(len(items) * 0.005)
        
        return [
            {'id': int(datetime.now().timestamp() * 1000) + i, **item, 'inserted': True}
            for i, item in enumerate(items)
        ]
    
    async def batch_update(self, updates: List[Dict[str, Any]]) -> List[Dict[str, Any]]:
        """Batch update"""
        if not updates:
            return []
        
        print(f"Batch updating {len(updates)} records...")
        await asyncio.sleep(len(updates) * 0.003)
        
        return [
            {**update, 'updated': True, 'updated_at': datetime.now().timestamp()}
            for update in updates
        ]
    
    async def batch_delete(self, ids: List[int]) -> List[Dict[str, Any]]:
        """Batch delete"""
        if not ids:
            return []
        
        print(f"Batch deleting {len(ids)} records...")
        await asyncio.sleep(len(ids) * 0.002)
        
        return [{'id': id, 'deleted': True} for id in ids]


async def demonstrate_batch_processing():
    """Demonstrate batch processing"""
    print('=== Batch Processing ===\n')
    
    # Batch processor
    processor = BatchProcessor({
        'batch_size': 5,
        'flush_interval': 3.0
    })
    
    # Add items
    print('Adding items to batch...')
    for i in range(1, 13):
        await processor.add({'id': i, 'data': f'item-{i}'})
    
    # Wait a bit
    await asyncio.sleep(1)
    print('Stats:', processor.get_stats())
    
    # Flush remaining
    await processor.flush()
    print('Final stats:', processor.get_stats())
    
    # Database batch operations
    print('\n=== Database Batch Operations ===\n')
    db_ops = DatabaseBatchOperations()
    
    # Batch insert
    items = [{'name': f'Item {i}', 'value': i * 10} for i in range(10)]
    inserted = await db_ops.batch_insert(items)
    print(f"Inserted {len(inserted)} records")
    
    # Batch update
    updates = [{'id': item['id'], 'name': f"Updated {item['name']}"} for item in inserted[:5]]
    updated = await db_ops.batch_update(updates)
    print(f"Updated {len(updated)} records")
    
    # Batch delete
    ids = [item['id'] for item in inserted[5:]]
    deleted = await db_ops.batch_delete(ids)
    print(f"Deleted {len(deleted)} records")
    
    await processor.shutdown()


if __name__ == '__main__':
    asyncio.run(demonstrate_batch_processing())

